File: worker/go/baseline.go
// Date         : 2018/12/16
// Author       : zhangchaoyang
// Description  : 
package main

import (
	"math"
	"time"
	"ParameterServer/util"
)

func fn(x []float64, w []float64) float64 {
	var z float64 = 0.0
	if len(x) != len(w) {
		panic("length not match")
	}
	for i := 0; i < len(x); i++ {
		z += x[i] * w[i]
	}
	//var y_hat float64 = 1.0 / (1.0 + math.Pow(math.E, -z))
	var y_hat float64 = (1.0 + math.Tanh(z/2.0)) / 2.0
	return y_hat
}

func crossEntropy(y float64, y_hat float64) float64 {
	l := -y*math.Log(y_hat+NEAR_0) - (1-y)*math.Log(1-y_hat+NEAR_0)
	return l
}

func batchGradient(sampleBatch []*Sample, w []float64) []float64 {
	gSum := make([]float64, len(w))
	for _, sample := range sampleBatch {
		y_hat := fn(sample.X, w)
		sub := y_hat - sample.Y
		//util.Log.Debugf("y=%f, y_hat=%f", sample.Y, y_hat)
		//只需要计算自己负责的区间内的导数
		for i := 0; i < len(w); i++ {
			v := sample.X[i]
			gSum[i] += sub * v //各维度上梯度求和
		}
	}
	
	g := make([]float64, len(w))
	for i, v := range gSum {
		g[i] = v / float64(len(sampleBatch)) //各维度上梯度求平均值
	}
	return g
}

func trainFTRL(ParameterTotal int, corpusFile string, epoch int, batch int, alpha float64, beta float64, l1 float64,
	l2 float64) []float64 {
	begin := time.Now()
	z := make([]float64, ParameterTotal)
	n := make([]float64, ParameterTotal)
	w := make([]float64, ParameterTotal)
	
	for ep := 0; ep < epoch; ep++ {
		util.Log.Infof("epoch=%d", ep)
		CorpusGenerator := new(CorpusGenerator)
		CorpusGenerator.Init(1, 0, corpusFile, ParameterTotal)
		sampleBatch := []*Sample{}
		for {
			sample := CorpusGenerator.PeekOneSample()
			if sample == nil {
				CorpusGenerator.Close()
				break
			}
			sampleBatch = append(sampleBatch, sample)
			if len(sampleBatch) >= batch {
				for i := 0; i < ParameterTotal; i++ {
					if math.Abs(z[i]) < l1 {
						w[i] = 0.0
					} else {
						l := l1
						if math.Signbit(z[i]) {
							l = -l1
						}
						w[i] = -(z[i] - l) / (l2 + (beta+math.Sqrt(n[i]))/alpha)
					}
				}
				g := batchGradient(sampleBatch, w)
				for i := 0; i < len(g); i++ {
					sigma := (math.Sqrt(n[i]+g[i]*g[i]) - math.Sqrt(
						n[i])) / alpha
					z[i] += g[i] - sigma*w[i]
					n[i] += g[i] * g[i]
				}
				sampleBatch = []*Sample{}
			}
		}
	}
	util.Log.Infof("train lr with ftrl finished, use %f seconds", time.Since(begin).Seconds())
	return w
}

func predictByFTRL(w []float64, corpusFile string) {
	begin := time.Now()
	ParameterTotal := len(w)
	CorpusGenerator := new(CorpusGenerator)
	CorpusGenerator.Init(10, 0, corpusFile, ParameterTotal)
	loss := 0.0
	population := 0
	for {
		sample := CorpusGenerator.PeekOneSample()
		if sample == nil {
			CorpusGenerator.Close()
			break
		}
		y_hat := fn(sample.X, w)
		loss += crossEntropy(sample.Y, y_hat)
		population++
	}
	util.Log.Infof("predit %d samples, mean loss %f, use %f seconds", population, loss/float64(population),
		time.Since(begin).Seconds())
}
func main3() {
	w := trainFTRL(10000, "../data/binary_class.csv", 20, 64, 0.1, 1.0, 1.0, 1.0)
	predictByFTRL(w, "../data/binary_class.csv")
	util.Log.Flush()
}

File: worker/python/baseline.py
# -*- coding:utf-8 -*-
# Author : zhangchaoyang
# Date : 2018/12/16 9:50 PM

from __future__ import division

import os
import sys
import time

sys.path.insert(0, os.getcwd())
import numpy as np

from util.logger import LoggerFactory

LoggerFactory.setDefaultLogger("log/ps_client.log", "INFO")
logger = LoggerFactory.getLogger()
from corpus import CorpusGenerator

NEAR_0 = 1e-10


class LR_CLS(object):
    '''用于二分类的LR
    '''

    @staticmethod
    def fn(w, x):
        '''决策函数为sigmoid函数。
           用tanh间接表示sigmoid，防止出现RuntimeWarning: overflowencountered in exp
        '''
        return (0.5 * (1 + np.tanh(0.5 * np.dot(x, w)))).reshape(x.shape[0], 1)

    @staticmethod
    def loss(y, y_hat):
        '''交叉熵损失函数'''
        return np.sum(
            np.nan_to_num(-y * np.log(y_hat + NEAR_0) - (1 - y) * np.log(1 - y_hat + NEAR_0)))  # 加上NEAR_0避免出现log(0)

    @staticmethod
    def grad(y, y_hat, x):
        '''交叉熵损失函数对权重w的一阶导数'''
        return np.mean((y_hat - y) * x, axis=0)


class LR_REG(object):
    '''用回归的LR
    '''

    @staticmethod
    def fn(w, x):
        '''决策函数为sigmoid函数'''
        return 1.0 / (1.0 + np.exp(-np.dot(x, w))).reshape(x.shape[0], 1)

    @staticmethod
    def loss(y, y_hat):
        '''误差平方和损失函数'''
        return np.sum(np.nan_to_num((y_hat - y) ** 2)) / 2

    @staticmethod
    def grad(y, y_hat, x):
        '''误差平方和损失函数对权重w的一阶导数'''
        return np.mean((y_hat - y) * (1 - y_hat) * y_hat * x, axis=0)


class FTRL(object):

    def __init__(self, dim, l1, l2, alpha, beta, decisionFunc=LR_CLS):
        self.dim = dim
        self.decisionFunc = decisionFunc
        self.z = np.zeros(dim)
        self.n = np.zeros(dim)
        self.w = np.zeros(dim)
        self.l1 = l1
        self.l2 = l2
        self.alpha = alpha
        self.beta = beta

    def predict(self, x):
        return self.decisionFunc.fn(self.w, x)

    def update(self, x, y):
        self.w = np.array([0 if np.abs(self.z[i]) <= self.l1 else (np.sign(self.z[i]) * self.l1 - self.z[i]) / (
                self.l2 + (self.beta + np.sqrt(self.n[i])) / self.alpha) for i in xrange(self.dim)])
        y_hat = self.predict(x)
        g = self.decisionFunc.grad(y, y_hat, x)
        sigma = (np.sqrt(self.n + g * g) - np.sqrt(self.n)) / self.alpha
        self.z += g - sigma * self.w
        self.n += g * g
        return self.decisionFunc.loss(y, y_hat)

    def train(self, corpus_file, verbos=False, epochs=100, batch=64):
        begin = time.time()
        total = 0
        for itr in xrange(epochs):
            logger.info("Epoch={:d}".format(itr))
            # 尽量使用mini batch，充分发挥numpy的并行计算能力
            mini_batch_x = []
            mini_batch_y = []
            n = 0
            corpus_generator = CorpusGenerator(corpus_file, 1, 0, self.dim)
            for X, Y in corpus_generator:
                n += 1
                mini_batch_x.append(X)
                mini_batch_y.append([Y])
                if len(mini_batch_x) >= batch:
                    self.update(np.array(mini_batch_x), np.array(mini_batch_y))
                    if verbos:
                        Y_HAT = self.predict(np.array(mini_batch_x))
                        train_loss = self.decisionFunc.loss(np.array(mini_batch_y), Y_HAT) / len(mini_batch_x)
                        logger.info("{:d}/{:d} train loss: {:f}\n".format(n, total, train_loss))
                    mini_batch_x = []
                    mini_batch_y = []
            self.update(np.array(mini_batch_x), np.array(mini_batch_y))
            if total == 0:
                total = n
            if verbos:
                Y_HAT = self.predict(np.array(mini_batch_x))
                train_loss = self.decisionFunc.loss(np.array(mini_batch_y), Y_HAT) / len(mini_batch_x)
                logger.info("{:d}/{:d} train loss: {:f}\n".format(n, total, train_loss))

        logger.info("train lr with ftrl finished, use {} seconds".format(time.time() - begin))

    def verify(self, corpus_file, xdim):
        begin = time.time()
        batch_x = []
        batch_y = []
        n = 0
        corpus_generator = CorpusGenerator(corpus_file, 10, 0, xdim)
        for X, Y in corpus_generator:
            n += 1
            batch_x.append(X)
            batch_y.append([Y])
        x = np.array(batch_x)
        y = np.array(batch_y)
        y_hat = self.predict(x)
        loss = self.decisionFunc.loss(y, y_hat) / x.shape[0]
        logger.info(
            "predit {} samples, mean loss {}, use {} seconds".format(n, loss, time.time() - begin))


if __name__ == "__main__":
    ftrl = FTRL(dim=10000, l1=1.0, l2=1.0, alpha=0.1, beta=1.0, decisionFunc=LR_CLS)
    ftrl.train("example/data/binary_class.csv", verbos=False, epochs=20, batch=64)
    ftrl.verify("example/data/binary_class.csv", xdim=10000)
